### ISSUE TITLE
[v2 BUG]: NaN target values set to 0.0 rather than being masked.

### ISSUE THREAD

**Issue by joelnkn** [chemprop/chemprop#960]
In MPNN evaluate_batch implementation, NaN target values are set to 0.0. This leads to unexpected metric results as these prediction, target pairs should be masked out.

https://github.com/chemprop/chemprop/blob/fc1d003091954b5e0e18cfcea94692cb3df0d591/chemprop/models/model.py#L177C4-L187C10

**Comment by KnathanM** [comment]
We took a look at this today in a metrics meeting and now think that the current implementation is correct. 

I see three places that we set NaN targets to zero:
https://github.com/chemprop/chemprop/blob/3213210c3519c7dc4e5ce452a57087f91b524115/chemprop/cli/train.py#L895-L896
https://github.com/chemprop/chemprop/blob/3213210c3519c7dc4e5ce452a57087f91b524115/chemprop/models/model.py#L143-L144
https://github.com/chemprop/chemprop/blob/3213210c3519c7dc4e5ce452a57087f91b524115/chemprop/models/model.py#L180-L181

In each of these cases, we first make a mask using `.isfinite` which returns `False` for NaN values (I previously thought it returned `True`). So the NaN targets are masked as desired. 

Why do we then set the NaN values to 0? In the loss functions, we multiply by the mask instead of indexing. False values are treated as 0, but 0 * NaN = NaN, so if the target was left as NaN, then the NaN would still get propagated forward. 
https://github.com/chemprop/chemprop/blob/3213210c3519c7dc4e5ce452a57087f91b524115/chemprop/nn/loss.py#L75

As a side note, [this](https://discuss.pytorch.org/t/performance-difference-of-mask-as-index-or-multiplication-with-mask/25668/2) post found that multiplying by the mask was significantly faster than indexing.
